src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java

```diff
@@ -0,0 +1,455 @@
+                                               @QueryParameter String mirrorName) {
+            return formValidation.doTestConnection(serverId, credentialsId, projectName, repositoryName, mirrorName);
+        }
+
+        @Nonnull
+        @Override
+        public String getDisplayName() {
+            return "Bitbucket server";
+        }
+
+        @Override
+        public List<GitSCMExtensionDescriptor> getExtensionDescriptors() {
+            return Collections.emptyList();
+        }
+
+        @Override
+        public List<GitTool> getGitTools() {
+            return Collections.emptyList();
+        }
+
+        public RetryingWebhookHandler getRetryingWebhookHandler() {
+            return retryingWebhookHandler;
+        }
+
+        @Override
+        public boolean getShowGitToolOptions() {
+            return false;
+        }
+
+        public List<SCMSourceTrait> getTraitsDefaults() {
+            return gitScmSourceDescriptor.getTraitsDefaults();
+        }
+
+        public List<NamedArrayList<? extends SCMSourceTraitDescriptor>> getTraitsDescriptorLists() {
+            return gitScmSourceDescriptor.getTraitsDescriptorLists();
+        }
+
+        @Override
+        protected SCMHeadCategory[] createCategories() {
+            return new SCMHeadCategory[]{UncategorizedSCMHeadCategory.DEFAULT, TagSCMHeadCategory.DEFAULT};
+        }
+
+        BitbucketScmHelper getBitbucketScmHelper(String bitbucketUrl,
+                                                 GlobalCredentialsProvider globalCredentialsProvider,
+                                                 @Nullable String credentialsId) {
+            return new BitbucketScmHelper(bitbucketUrl,
+                    bitbucketClientFactoryProvider,
+                    globalCredentialsProvider,
+                    credentialsId, jenkinsToBitbucketCredentials);
+        }
+
+        Optional<BitbucketServerConfiguration> getConfiguration(@Nullable String serverId) {
+            return bitbucketPluginConfiguration.getServerById(serverId);
+        }
+
+        private BitbucketMirrorHandler createMirrorHandler(BitbucketScmHelper helper) {
+            return new BitbucketMirrorHandler(
+                    bitbucketClientFactoryProvider,
+                    jenkinsToBitbucketCredentials,
+                    (client, project, repo) -> helper.getRepository(project, repo));
+        }
+    }
+
+    /**
+     * This class exists to work around two issues
+     * 1. We do not want to re-implement the retrieve found in the GitSCMSource, however it is protected so we can't
+     * access it from our class. This class inherits from the GitSCMSource and thus can access it and expose a method
+     * wrapper.
+     * 2. We want to inject a custom extension to enable posting build statues. The enforcer will not allow us to access
+     * it directly, so we override the getExtensions method and inject our own extension there.
```

Comment:
This makes me sad.

Reply:
yeah, one could argue we are doing the wrong thing though. It is definitely a hack but not too bad a hack I think.